src/NodaTime/Duration.cs

```diff
@@ -172,6 +172,41 @@ private Duration(long units, string paramName, long minValue, long maxValue, lon
             }
         }
 
+        // Note: unitsPerDay is always actually an integer, but we always want to perform floating point arithmetic
+        // with it.
+        private Duration(double units, string paramName, double unitsPerDay, long nanosPerUnit)
+        {
+            // Note: we don't check the original double values other than for NaN and infinity.
+            // Instead, we do the relevant conversion and then check that we're in bounds.
+            // That way we won't get invalid values due to weird rounding issues.
+            if (double.IsNaN(units) || double.IsInfinity(units))
+            {
+                throw new ArgumentOutOfRangeException(paramName, "Value must be a finite, valid number.");
+            }
+            unchecked
+            {
+                days = (int)(units / unitsPerDay);
+                double unitOfDay = units - (unitsPerDay * days);
+                if (unitOfDay < 0)
+                {
+                    days--;
+                    unitOfDay += unitsPerDay;
+                }
+                nanoOfDay = (long) (unitOfDay * nanosPerUnit + 0.5);
+
+                // Just in case the rounding up tips us into a new day...
```

Comment:
`// Just in case the rounding up tips us into a new day...`

Would it be possible to avoid this block by doing the conversion to `nanoOfDay` (with rounding up) first and then accounting for negative values?

Reply:
Hmm. Possibly - will need to try it.